Add idempotency check to PostgreSQL migration test

juliogonzalez opened the pull request:
## What does this PR change?

Add idempotency check to PorgreSQL migration test

Examples:
- non idempotent change (migration failed): https://github.com/uyuni-project/uyuni/pull/156
- non idempotent change (migration passes, diff does not): https://github.com/uyuni-project/uyuni/pull/162
- idempotent change with separate oracle/postgresql scripts: https://github.com/uyuni-project/uyuni/pull/157
- idempotent change with single script: https://github.com/uyuni-project/uyuni/pull/174

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: I'd suggest a Wiki page at the Wiki for this repository with some ideas about how to write idempotent SQL scripts. @moio what do you think? Could you take care of that?

- [ ] **DONE**

## Test coverage
- No tests: This is a test itself.

- [x] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/5819

- [x] **DONE**

juliogonzalez commented:
gitarro no changelog needed !!!

juliogonzalez commented:
gitarro rerun changelog_test

juliogonzalez commented:
gitarro rerun changelog_test !!!

juliogonzalez commented:
gitarro rerun changelog_test !!!

juliogonzalez commented:
gitarro rerun changelog_test !!!

juliogonzalez commented:
gitarro rerun changelog_test !!

juliogonzalez commented:
gitarro rerun changelog_test !!!

juliogonzalez commented:
gitarro rerun changelog_test !!!

juliogonzalez commented:
I am adding three reviewers here, as I consider this test to be specially important:

- @mcalmer as python developer and because of his experience
- @moio as our SQL expert
- @MalloZup as QA (the rest of the team is too busy with retail.

MalloZup commented:
As open question: instead of.putting it to spacewalk, we could have the python script in the susemanager ci repo with pipelines

juliogonzalez commented:
@MalloZup I agree, but I think as a second step, outside this PR.

We should keep in mind:
- Moving this script means moving the migration script as well
- We should publish the containers to run it (right now the stuff to generate it as this repo, and containers are inside a private registry).
- Maybe there could be other dependencies.

MalloZup commented on susemanager-utils/testing/automation/schema-migration-test-pgsql.sh:
```diff
@@ -4,6 +4,23 @@ HERE=`dirname $0`
+
+# Base version for the idempotency test. All PostgreSQL since this version must be idempotent for the test to pass
+# This value is is used if the file for the PR is not found (we assume we are on a branch)
+BASE_VERSION='4.0.1'
```
we might put the base_version in the https://github.com/uyuni-project/uyuni/blob/1c0c04d036b7c2693585140c8a49d388533ab2c8/susemanager-utils/testing/automation/VERSION file?

From external POV we might call this var `BASE_VERSION` like `PGSQL_BASE_VERSION` if it has impact for base_version.

juliogonzalez replied:
Fixed and testing.

MalloZup commented on susemanager-utils/testing/automation/schema-migration-test-pgsql.sh:
```diff
@@ -4,6 +4,23 @@ HERE=`dirname $0`
+  IDEMPOTENCY_PARAMS=" -v ${BASE_VERSION}"  
+fi
+
+docker run --privileged --rm=true -v "$GITROOT:/manager" $REGISTRY/$PGSQL_CONTAINER /bin/bash -c "/manager/susemanager-utils/testing/docker/scripts/schema_migration_test_pgsql-31to40.sh && /manager/susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py ${IDEMPOTENCY_PARAMS}"
```
we could grap the Scripts in variable
like this one
`/manager/susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py`

I think this var can stay in this script since is local to this one

juliogonzalez replied:
Fixed and testing.

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/python
```
can we make this python3 specific ?

juliogonzalez replied:
Fixed and testing.

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+        open(path, 'r')
+        return(True)
+    except FileNotFoundError:
+        return(False)
```
` Unnecessary parens after 'return' keyword (superfluous-parens)`

MalloZup replied:
same in all return from pylint

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    for element in os.listdir(schemaPath):
+        elementVer = re.search(
+            '^susemanager-schema-(\d+\.\d+[\d|.]*)-to-susemanager-schema-\d+\.\d+[\d|.]*$', element)
+        if elementVer:
```
i would use here GuardeClause.

`if ! elementVer: continue`

So you gain simplity of code ( nested if add also complexity) and is difficult to read

juliogonzalez replied:
Fixed and testing. One of this days I will be able to remember about it :-D

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    for element in os.listdir(schemaPath):
+        elementVer = re.search(
+            '^susemanager-schema-\d+\.\d+[\d|.]*-to-susemanager-schema-(\d+\.\d+[\d|.]*)$', element)
+        if elementVer:
```
i would use personally guarde clause

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    os.mkdir(fakePath)
+    for migrationFile in files:
+        shutil.copy(migrationFile, fakePath)
+    #run_command(sys.path[0] + "/build-schema.sh")
```
if the command is commented out i would not commit it

juliogonzalez replied:
Ooops. This was part of a test. Good catch.

Removed.

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+                        help="Path where the spacewalk-schema-upgrade script is (default: /manager/schema/spacewalk/spacewalk-schema-upgrade)",
+                        default="/manager/schema/spacewalk/spacewalk-schema-upgrade")
+    args = parser.parse_args()
+    if args.schemaPath is None:
```
i would create a `argparser` helper function for extracting this from main

juliogonzalez replied:
I will give it a try tomorrow.

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    print("                     Idempotency test                       ")
+    print("============================================================")
+    # Exclude tables and sequences that are not updated by the SQL scripts (version control)
+    excludedTables = ["pg_catalog.setval", "public.rhnpackageevr",
```
i would personally put them outside of main and pass them as ARG to main. 

If python has constant like golang/ruby i would freeze them

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    for line in diff_dumps(initial_dump, migrated_dump):
+        diff = True
+        print(line)
+    if diff:
```
i would vote for GuardClause `if ! diff contiune`

juliogonzalez replied:
Done and testing.

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+def get_all_files_rpm_pr(prFile):
+    with open(prFile, 'r') as f:
+        changedFiles = []
+        files = json.loads(read(f))['files']
```
i would personally split this function for more readablity:

` with open(prFile, 'r') as f:
        changedFiles = []
        files = json.loads(read(f))['files']
`

` with open(prFile, 'r') as f:
        changedFiles = []
        list_of_changed_files_by_pr  = read(f))['files'] // i dunno if is the python way for calling vars
        files = json.loads(list_of_changed_files_by_pr)
`

juliogonzalez replied:
I added @moio suggestion, so the the function is now shorter.

MalloZup commented:
@juliogonzalez regarding the moving-out/moving in i opened an issue https://github.com/uyuni-project/uyuni/issues/165

moio commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+import sys
+
+
+def file_exists(path):
```
Is there a particular reason not to use `os.path.isfile(path)`?

juliogonzalez replied:
To be honest, I don't know what I was thinking about x'D

Fixed and testing.

moio commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+        return(False)
+
+
+def directory_exists(path):
```
I would suggest to simply inline `os.path.isdir(path)` and remove this function.

If you are tired of typing dots (which is something I can fully understand) you can also import the isdir/isfile symbols directly.

juliogonzalez replied:
Same as before.

Fixed and testing.

moio commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+        files = json.loads(read(f))['files']
+        for file in filesDict:
+            if file['status'] == 'modified' or file['status'] == 'added':
+                changedFiles.append(file['filename'])
```
The currently-in-fashion way to do this kind of collection computation is by using list comprehensions (functional programming). In this case it is something like:

```python
changedFiles = [file['filename'] for file in filesDict if file['status'] == 'modified' or file['status'] == 'added']
```

juliogonzalez replied:
Fixed and testing.

juliogonzalez replied:
In this case I realized I did not add the regex to look if the file is a SQL file from the migration scripts.

So now the line looks:
```
changed_files = [file['filename'] for file in files if re.search('^schema/spacewalk/upgrade/susemanager-schema-\d+\.\d+[\d|.]*-to-susemanager-schema-\d+\.\d+[\d|.]*/.+$', file['filename']) and (file['status'] == 'modified' or file['status'] == 'added'])
```

And while I agree this in fashion, I would say it's harder to read than:
```
changed_files = [ ]
for file in files:
    if re.search('^schema/spacewalk/upgrade/susemanager-schema-\d+\.\d+[\d|.]*-to-susemanager-schema-\d+\.\d+[\d|.]*/.+$', file['filename']) and (file['status'] == 'modified' or file['status'] == 'added']):
        changed_files.append(file['filename'])
```
Thoughts?

MalloZup replied:
:+1:  for the readable one ake 2nd

moio commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    return(lastVersion.rsplit('.', 1)[0] + '.' + str(int(lastVersion.rsplit('.', 1)[1])+1))
+
+
+def create_fake_migration_path(schemaPath, newVersion, prFile=None, version=None):
```
Python coders attempt to distinguish themselves from Java coders by favoring `short_names` instead of `camelCasedNamesEspeciallyVeryVerboseButToThePointOnes`.

juliogonzalez replied:
Are we talking about the variables here? The function name is already snake cased (not short, but I don't really know how to make it shorter).

About the variables: yes, I will need to change them to snake case.

moio commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+        if elementVer:
+            # StrictVersion does not work as we have some migrations with version X.Y.W.Z
+            # https://hg.python.org/cpython/file/tip/Lib/distutils/version.py#l93
+            if DV.LooseVersion(elementVer.group(1)) >= DV.LooseVersion(version):
```
Oooh comparing versions is a rabbit hole.

https://blog.jasonantman.com/2014/07/how-yum-and-rpm-compare-versions/

I would suggest sticking to the official implementation, `rpmvercmp`: https://fedoraproject.org/wiki/Archive:Tools/RPM/VersionComparison

There are Python bindings in the rpmdevtools package AFAIU. https://utcc.utoronto.ca/~cks/space/blog/linux/RPMShellVersionComparison

juliogonzalez replied:
Well, strictly speaking this is about the schema versions at the directories, not RPM versions (no epoch, no release, no dist...).

That's the reason this python function works so well to just compare X.Y.W[.Z].

I will review what else can be done, but the idea was sticking to the Python standard library instead of adding more stuff to the containers (otherwise I'd be using Yaml -as @MalloZup requested instead of JSON).

moio commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    actions = ['start', 'stop']
+    if action not in actions:
+        raise "Invalid action for the PostgreSQL service"
+    if file_exists("/usr/lib/postgresql-init"):
```
In what case does this not exist? Can't we always use `rcpostgresql`?

juliogonzalez replied:
This is basically a copy paste of what we do at `schema_migration_test_pgsql-31to40.sh` (we use bash there), so I guess there is a good reason... or maybe just a historic one :-D

@mcalmer?

mcalmer replied:
Maybe this was added with postgres 9.6 and the old rcpostgresql script do not work anymore.
So if this script exists, use it, otherwise try rcpostgresql command.

If you work for Uyuni only you can use /usr/lib/postgresql-init exclusivly

juliogonzalez replied:
Well, in fact the base OS used at the container is SLE12 using PostgreSQL 9.6.

So this means we can safely use `/usr/lib/postgresql-init` only from now on, as both Uyuni and future SUSE Manager will use >= 9.6.

Thanks for the clarification @mcalmer, I will fix both tests.

moio commented on susemanager-utils/testing/docker/scripts/schema_idempotency_test_pgsql.py:
```diff
@@ -0,0 +1,225 @@
+    print("============================================================")
+    # Exclude tables and sequences that are not updated by the SQL scripts (version control)
+    excludedTables = ["pg_catalog.setval", "public.rhnpackageevr",
+                      "public.rhn_pkg_evr_seq", "public.rhnversioninfo"]
```
I am not sure if a whitelist approach will work in the long run but let's try and see how long the list becomes as we discover false positives.

juliogonzalez replied:
This is the absolute minimum, as the sequences and tables to control schema versions are always going to change.

If we want to add further ignores, I'd add a new argument and a file.

juliogonzalez commented:
Almost all commets addressed, and the linter looks much better now.

I will try to create an argparse tomorrow, and will test everything again, to get a new review.

Thanks everyone for your comments so far!

MalloZup commented:
Awesome Julio, feel free to ask me any review for your passion I could do anything but not washing your car! Thx for yur spirit always a pleasure

MalloZup commented on susemanager-utils/testing/docker/scripts/schema_migration_test_pgsql-31to40.sh:
```diff
@@ -58,7 +62,7 @@ fi
 ################################################
 
-set -x
+# set -x
```
if commented out we can remove

juliogonzalez commented:
Ok, with a couple of exceptions, all required changes are implemented.

I also added some simple docstrings to the functions, and fixed several problems detected by the linter, so things look much better now:

```
************* Module schema_idempotency_test_pgsql
C: 19, 0: Line too long (142/100) (line-too-long)
C:161, 0: Line too long (131/100) (line-too-long)
C:164, 0: Line too long (181/100) (line-too-long)
C:168, 0: Line too long (143/100) (line-too-long)
C:225, 0: Line too long (147/100) (line-too-long)
C:  1, 0: Missing module docstring (missing-docstring)

------------------------------------------------------------------
Your code has been rated at 9.63/10 (previous run: 9.63/10, +0.00)
```

MalloZup commented:
@juliogonzalez i think if look at spacewalk, we don\t have such ratings from pylint .. ` 9.63/10` :+1:  :smile_cat:

juliogonzalez commented:
All commits squashed. I will fix the PRs to test with the same script, and if everything is fine I will merge.
